spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala

```diff
@@ -475,30 +253,139 @@ class Graph[T: ClassTag](val inputs : Seq[ModuleNode[T]],
    * @return current graph model
    */
   def stopGradient(names: Array[String]): this.type = {
+    if (stopGradientLayers == null) stopGradientLayers = new util.HashSet[String]()
+
     names.foreach(name => {
       val layer = this (name)
       require(layer.isDefined, s"cannot find layer match ${name}")
-      if (stopGradientLayers == null) stopGradientLayers =
-        new util.HashSet[String]()
       stopGradientLayers.add(layer.get.getName())
     })
-    build()
+    buildBackwardGraph()
+    this
+  }
+
+  /**
+   * set an array of layers that match the given ```names``` to be "freezed",
+   * i.e. their parameters(weight/bias, if exists) are not changed in training process
+   * @param names an array of layer names
+   * @return current graph model
+   */
+  def freeze(names: Array[String]): this.type = {
+    names.foreach(name => {
+      val layer = this (name)
+      require(layer.isDefined, s"cannot find layer match ${name}")
+      layer.get.setScaleW(0)
+      layer.get.setScaleB(0)
+    })
     this
   }
 
+  private[bigdl] def removeStopNodes(n: Node[_]): Unit = {
+    val nodes = n.nextNodes
+    n.removeNextEdges()
+    nodes.filter(_.prevNodes.length == 0).foreach(removeStopNodes(_))
+  }
+
+  protected val inputCache = new mutable.HashMap[String, Activity]()
+  protected val gradOutputCache = new mutable.HashMap[String, Activity]()
+
+  protected def getInput(
+    node: Node[AbstractModule[Activity, Activity, T]],
+    input: Activity
+  ): Activity = {
+    if (inputs.length == 1) {
+      require(inputs(0).eq(node), "input node is not in the input list")
+      input.toTensor
+    } else {
+      val i = inputs.indexOf(node)
+      require(i != -1, "input node is not in the input list")
+      input.toTable[Tensor[T]](i + 1)
+    }
+  }
+
+  protected def findInput(node: ModuleNode[T], input: Activity): Activity = {
+    val nodeInput = if (node.prevNodes.isEmpty && !node.element.isInstanceOf[WithoutInput]) {
+      getInput(node, input)
+    } else {
+      val prevActivities = node.prevNodesAndEdges
+        .filterNot(n => n._1.element.isInstanceOf[ControlDependency[T]])
+        .map(n => {
+          n._2.fromIndex match {
+            case Some(i) =>
+              if (n._1.element.output == null || (i == 1 && n._1.element.output.isTensor)) {
+                n._1.element.output
+              } else {
+                n._1.element.output.toTable.apply[Activity](i)
+              }
+            case None => n._1.element.output
+          }
+        })
+      if (prevActivities.length == 1) {
```

Comment:
If prevActivities is empty, an empty table will be returned?

Reply:
If preActivities is empty, it will not go to this "else" branch

Reply:
What if (node.prevNodes.isEmpty && node.element.isInstanceOf[WithoutInput])?

Reply:
Such node const/variable won't use the input...

Reply:
fixed this